R/reactives.R

```diff
@@ -449,6 +450,38 @@ reactive <- function(x, env = parent.frame(), quoted = FALSE, label = NULL,
   structure(o$getValue, observable = o, class = "reactive")
 }
 
+# Given the srcref to a reactive expression, attempts to figure out what the
+# name of the reactive expression is. This isn't foolproof, as it literally
+# scans the line of code that started the reactive block and looks for something
+# that looks like assignment. If we fail, fall back to a default value (likely
+# the block of code in the body of the reactive).
+srcrefToLabel <- function(srcref, defaultLabel) {
+  if (is.null(srcref))
+    return(defaultLabel)
+
+  srcfile <- attr(srcref, "srcfile", exact = TRUE)
+  if (is.null(srcfile))
+    return(defaultLabel)
+
+  if (is.null(srcfile$lines))
+    return(defaultLabel)
+
+  firstLine <- substring(srcfile$lines[srcref[1]], 1, srcref[2] - 1)
+  m <- regexec("(.*)(<-|=)\\s*reactive\\s*\\($", firstLine)
+  if (m[[1]][1] == -1) {
+    return(defaultLabel)
+  }
+  sym <- regmatches(firstLine, m)[[1]][2]
+  res <- try(parse(text = sym), silent = TRUE)
+  if (inherits(res, "try-error"))
+    return(defaultLabel)
+
+  if (length(res) != 1 || !is.symbol(res[[1]]))
```

Comment:
I'm wrong, `a = list(foo` won't parse. The (even more unrealistic) case I was originally thinking of was `a <- 1; b <- eactive(...)`. The length(res) check would take care of that, I could drop the `is.symbol` test, I think that would satisfy your case.